src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteOutputService.java

```diff
@@ -0,0 +1,527 @@
+
+      blockingStub.finalizeBuild(builder.build());
+
+      stateAcrossBuilds.previousBuildId = stateAcrossBuilds.currentBuildId;
+      stateAcrossBuilds.currentBuildId = null;
+      currentOriginalOutputPath = null;
+      currentRelativeOutputPath = null;
+    }
+  }
+
+  @Override
+  public void finalizeAction(Action action, MetadataHandler metadataHandler)
+      throws IOException, EnvironmentalExecException, InterruptedException {
+    // TODO: Would this be the right place to call into the remote
+    // output service to check whether any I/O errors occurred? If so,
+    // we should likely let createActionFileSystem() call into the
+    // remote output service to start capturing I/O errors.
+  }
+
+  private String fixupExecRootPath(PathFragment path) {
+    return path.relativeTo(currentRelativeOutputPath).toString();
+  }
+
+  private class GrpcBatchStat implements BatchStat {
+    private abstract class DumbFileStatus implements FileStatusWithDigest {
+      @Override
+      public long getSize() throws IOException {
+        throw new RuntimeException("Not implemented");
+      }
+
+      @Override
+      public long getLastModifiedTime() throws IOException {
+        throw new RuntimeException("Not implemented");
+      }
+
+      @Override
+      public long getLastChangeTime() throws IOException {
+        return -1;
+      }
+
+      @Override
+      public long getNodeId() throws IOException {
+        throw new RuntimeException("Not implemented");
+      }
+
+      @Override
+      public byte[] getDigest() throws IOException {
+        throw new RuntimeException("Not implemented");
+      }
+    }
+
+    private class RegularFileStatus extends DumbFileStatus {
+      private final long size;
+      private final byte[] digest;
+
+      RegularFileStatus(long size, byte[] digest) {
+        this.size = size;
+        this.digest = digest;
+      }
+
+      @Override
+      public boolean isFile() {
+        return true;
+      }
+
+      @Override
+      public boolean isDirectory() {
+        return false;
+      }
+
+      @Override
+      public boolean isSymbolicLink() {
+        return false;
+      }
+
+      @Override
+      public boolean isSpecialFile() {
+        return false;
+      }
+
+      @Override
+      public long getSize() throws IOException {
+        return size;
+      }
+
+      @Override
+      public byte[] getDigest() throws IOException {
+        return digest;
+      }
+    }
+
+    private class SymlinkFileStatus extends DumbFileStatus {
+      @Override
+      public boolean isFile() {
+        return false;
+      }
+
+      @Override
+      public boolean isDirectory() {
+        return false;
+      }
+
+      @Override
+      public boolean isSymbolicLink() {
+        return true;
+      }
+
+      @Override
+      public boolean isSpecialFile() {
+        return false;
+      }
+    }
+
+    public List<FileStatusWithDigest> batchStat(boolean includeDigest,
```

Comment:
This is a nice optimization. One thing I would suggest is to consider having that in a follow up PR, especially since there seem to be some TODOs -- this way we wouldn't need to block the main part of this change on getting all that done.

Reply:
As most of the of the TODOs have already been processed now, this functionality seems to work quite reliable now. Would you mind if immediately we aimed for getting it included?